jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java

```diff
@@ -0,0 +1,735 @@
+        double z = Double.parseDouble(elements[3]);
+        return vectorFactory.apply(new double[] { x, y, z });
+    }
+
+    private static String[] splitLineElements(String line, int expectedLineElements)
+    {
+        String[] elements = line.split(" ");
+        if (elements.length != expectedLineElements) {
+            throw new IllegalArgumentException(
+                "Unexpected number of line elements: " + elements.length + " in line: " + line);
+        }
+        return elements;
+    }
+
+    // read of coordinates and graph computation
+
+    /**
+     * Read the complete {@link Graph} of node locations and returns true if all locations are
+     * distinct.
+     */
+    private boolean readCompleteNodeGraphData(
+        Graph<V, DefaultWeightedEdge> graph, BufferedReader reader,
+        Function<String, V> vectorFactory, ToIntBiFunction<V, V> edgeWeightFunction,
+        Integer dimension)
+        throws IOException
+    {
+        List<V> nodeCoordinates = readNodeCoordinateData(reader, vectorFactory, dimension);
+
+        buildCompleteGraph(graph, nodeCoordinates, edgeWeightFunction);
+        return graph.vertexSet().size() == nodeCoordinates.size();
+    }
+
+    private List<V> readNodeCoordinateData(
+        BufferedReader reader, Function<String, V> vectorFactory, Integer dimension)
+        throws IOException
+    {
+        List<V> coordinates = dimension != null ? new ArrayList<>(dimension) : new ArrayList<>();
+
+        for (String line; (line = reader.readLine()) != null;) {
+            String lineContent = line.trim();
+            if ("EOF".equals(lineContent)) {
+                break;
+            }
+            coordinates.add(vectorFactory.apply(lineContent));
+        }
+        return coordinates;
+    }
+
+    private static <V, E> void buildCompleteGraph(
+        Graph<V, E> graph, List<V> locations, ToIntBiFunction<V, V> edgeWeightFunction)
+    {
+        locations.forEach(graph::addVertex);
+
+        new CompleteGraphGenerator<V, E>().generateGraph(graph, null);
+
+        // Compute edge weights:
+        // Enable parallel streams. Synchronization not necessary, because DefaultWeightedEdge are
+        // IntrusiveEdges and modified directly.
+
+        graph.edgeSet().parallelStream().forEach(e -> {
+            V s = graph.getEdgeSource(e);
+            V t = graph.getEdgeTarget(e);
+            double weight = edgeWeightFunction.applyAsInt(s, t);
+            graph.setEdgeWeight(e, weight);
+        });
+    }
+
+    private List<Integer> readTourData(BufferedReader reader, Integer dimension)
+        throws IOException
+    {
+        List<Integer> tour = dimension != null ? new ArrayList<>(dimension) : new ArrayList<>();
+
+        for (String line; (line = reader.readLine()) != null;) {
+            String lineContent = line.trim();
+            if ("-1".equals(lineContent)) {
+                break;
+            }
+            tour.add(Integer.valueOf(lineContent));
```

Comment:
You need to consult some `Map<Integer, V>` here which you probably need to build earlier in your code in order to create a tour which is a `List<V>`.